### ISSUE TITLE
Add RNN and LSTM modules in `nn`

### ISSUE THREAD

**Issue by chelsea0x3b** [coreylowman/dfdx#204]
Should have tuple input and outputs with hidden state.

**Comment by Dimev** [comment]
Still kinda confused on how training would happen/how the gradients would be recorded.

Afaik the other deep learning libraries expand it into a higher dimension tensor that contains the output for each step, so you'd need a fold somewhere.

**Comment by coreylowman** [comment]
Gradient tracking wouldn't work differently than the other stuff in nn. The inputs would just be different. For example pytorch outlines a couple of cases in their lstm docs https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html:
1. Give the input (N, L, H_in) and no hidden states -> the hidden states are initialized to zero and after running the lstm they are returned
2. Give the input (N, L, H_in) and hidden states -> use hidden states as passed in and still return the new ones after running the lstm
3. Both of the above for unbatched input (without the first dimension)

for dfdx this would just be multiple `impl Module<...>`. pytorch and dfdx would both require that all items in the batch have the same sequence length.

**Comment by TimerErTim** [comment]
I am not that familiar with RNNs but is it really necessary to manually provide the hidden state and return the new one as output?

Wouldn't it just work if the Module stored the hidden state of its previous invocation internally and then used it again as its input upon the next invocation? Seems easier and more intuitive for users.  

The hidden and carry state (in case of an LSTM) could optionally be returned in the output, but I don't think it is good to require the hidden state in the input unless strictly necessary by nature.

**Comment by hsn10** [comment]
LSTM is one of few major module types still missing
